### PR TITLE
Initialize optimizer in dynamo to avoid graph break and tracing slowness

### PR DESCRIPTION
**Pull request by mlazos**
On calls to `_init_group` rather than tracing through it, extract python values from the arguments, and call the initialization. This avoids having to trace this function which is very slow with large parameters, and also avoids graph breaking on it. This is sound in this case because the state is only initialized once in the eager case. Guards on the state and params are generated explicitly rather than via tracing the initialization. 

Caveats:
`_init_group` also gathers various state tensors into lists via mutating list arguments to pass to the functional optimizer implementation. These state tensors exist on the optimizer itself, but we don't know exactly how the gathering is done and which tensors correspond to which attributes of the optimizer module (each optimizer has different states). To rectify this, we keep weak_ptrs to all of the tensors collected in the lists in globals (similar to how parameter keys are stored for dictionaries). These pointers are guaranteed to be alive as long as the optimizer object is alive if the internal state is not interfered with and they are guarded with weakref guards 

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy

### CODE REVIEW

File: torch/_dynamo/variables/optimizer.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    def get_python_args(self, *args, **kwargs):
+        """Get python values equivalent to the variable tracker args"""
+
+        def map_arg(arg):
+            if isinstance(arg, ConstantVariable):
+                return arg.as_python_constant()
+            elif isinstance(arg, ListVariable) and not arg.items:
+                return []
+            elif isinstance(arg, ConstDictVariable) and isinstance(
+                arg.source, GetItemSource
```
Comment (jansel) [inline torch/_dynamo/variables/optimizer.py:56]: I think we need to check that arg.source is actually pointing to the param_groups here.

**Review by jansel — APPROVED** [review]
Overall this approach seems reasonable to me.  Made one small comment about correct.

**Comment by eellison** [comment]
There was a 10s HF compilation latency increase between 8215468870 and 87cbfe957a. I think this is the only plausible culprit.

**Comment by janeyx99** [comment]
the optim benchmarks also started running into bugs after this change, see: https://github.com/pytorch/benchmark/actions/runs/5167132765/jobs/9307817625

can we revert or back out this change, add some tests and verify the bugs no longer exist, and then reland?

**Comment by ezyang** [comment]
see also https://github.com/pytorch/test-infra/issues/4282

**Comment by janeyx99** [comment]
Sigh, I see backing out is not trivial because of https://github.com/pytorch/pytorch/pull/103121.

**Pull request closed without merging**